fleet_notifications/state_checker.py

```diff
@@ -1,110 +1,152 @@
+                self.orders[state.order_id] = self.order_api.get_order(car_id=car_id, order_id=state.order_id)
+            except:
+                logger.error(f"Error while getting order with ID {state.order_id} from the api.")
+                return False
+        new_order_started = False
+        for order in self.orders.values():
+            if order.car_id == car_id:
+                new_order_started = True
+                break
+
+        if (no_active_order and new_order_started and self.orders[state.order_id].last_state.status
+            != OrderStatus.DONE and self.orders[state.order_id].last_state.status != OrderStatus.CANCELED):
+            logger.info(f"New mission started.")
+            threading.Thread(target=self.notification_client.call_phone, daemon=True,
+                             args=(admin_phone, under_test,)).start()
+        return True
+
+
+    def _check_if_order_is_done(self, car_id: int, state: OrderState, under_test: bool) -> None:
+        """Checks if the order is newly done and sends a notification to the phone number in the order."""
+        if self.orders[state.order_id].last_state.status != OrderStatus.DONE and state.status == OrderStatus.DONE:
+            logger.info(f"Order {state.order_id} is done.")
+            self.orders[state.order_id] = self.order_api.get_order(car_id=car_id, order_id=state.order_id)
+
+            notification_phone = self.orders[state.order_id].notification_phone
+            if(notification_phone is None):
+                logger.warning(f"Order {state.order_id} has no notification phone number.")
+                return
+
+            threading.Thread(target=self.notification_client.call_phone, daemon=True, args=(
+                             notification_phone.phone, under_test,)).start()
+
+
+    def _remove_finished_orders(self, since: int) -> None:
+        """Removes finished orders from the list and the database. If an order is not finished,
+        the since parameter is used as its latest timestamp."""
+        orders_to_remove = []
+        for order in self.orders.values():
+            if order.last_state.status == OrderStatus.DONE or order.last_state.status == OrderStatus.CANCELED:
+                orders_to_remove.append(order.id)
+            else:
                 try:
-                    car = car_api.get_car(car_id)
-                    car_id = car.id
-                    admin_phone = car.car_admin_phone.phone
-                    under_test = car.under_test
+                    self.order_api.get_order(car_id=order.car_id, order_id=order.id)
+                    notifications_db.update_order(order.id, order.car_id, since)
                 except:
-                    logging.error(f"Car not found: {car_id}")
+                    orders_to_remove.append(order.id)
+        for order_id in orders_to_remove:
+            self.orders.pop(order_id)
+            notifications_db.delete_order(order_id)
+
+
+    def _start(self) -> None:
```

Comment:
I suggest to extract the for loop as a non-public method and also simplifying the for loop itself.

``` suggestion
    def _start(self) -> None:
        """Starts checking order states on the Fleet Management API and triggers notifications when needed.
        This function runs indefinitely and should be run in a separate thread. If an error occurs,
        the function will sleep for a few seconds and then restart."""
        since = self._load_unfinished_orders() + 1
        while True:
            try:
                states: dict[int, OrderState] = {
                    state.order_id: state
                    for state in self.order_state_api.get_all_order_states(wait=True, since=since)
                }
                if states:
                    since = max(states.values(), key=lambda state: state.timestamp).timestamp
                    self._check_orders_and_call_if_done(states)
                    self._remove_finished_orders()
                    self._update_latest_timestamps(since)

            except KeyboardInterrupt:
                logger.info("Exiting the script.")
                return
            except Exception as e:
                logger.error(f"Unknown error: {e}, restarting.")
                time.sleep(THREAD_RESTART_DELAY)

    def _check_orders_and_call_if_done(self, new_states: dict[int, OrderState]) -> None:
        """Checks if the orders in the new states are new or done and triggers notifications if needed.

        `new_states` is a dictionary with order IDs as keys and the new states (with corresponding order ID) as values.
        """
        all_orders = {order.id: order for order in self.order_api.get_orders()}
        for order_id, state in new_states.items():
            logger.info(
                f"New order state ID: {state.id} for order {order_id} with status {state.status.name}"
            )
            order = all_orders.get(order_id, None)
            if not order:
                logger.error(f"Order not found: {order_id}")
                continue
            try:
                car = self.car_api.get_car(order.car_id)
            except:
                logger.error(f"Car not found: {order.car_id}")
                continue
            phone = "" if car.car_admin_phone is None else car.car_admin_phone
            if self._check_if_order_is_new(car.id, state, phone, car.under_test):
                self._check_if_order_is_done(car.id, state, car.under_test)
```